File: MobileBank/services/dashboardService.ts
import { dashboardApi } from './api';
import { API_CONFIG } from '@/config/environment';
import { AccountDTO } from './accountService';
import { ClientDTO } from './clientService';
import { TransactionDTO } from './transactionService';

// Types
export interface DashboardResponse {
  client: ClientDTO;
  accounts: AccountDTO[];
  recentTransactions: TransactionDTO[];
}

export interface StatementResponse {
  account: AccountDTO;
  startDate: string;
  endDate: string;
  transactions: TransactionDTO[];
}

// Types enrichis calculés côté client
export interface EnrichedDashboardResponse extends DashboardResponse {
  totalBalance: number;
  monthlyIncome: number;
  monthlyExpenses: number;
}

export interface EnrichedStatementResponse extends StatementResponse {
  totalCredits: number;
  totalDebits: number;
  balance: number;
}

const ENDPOINTS = API_CONFIG.DASHBOARD_SERVICE.ENDPOINTS;

export const dashboardService = {
  /**
   * Récupération du dashboard complet d'un client
   */
  getDashboard: async (clientId: number): Promise<DashboardResponse> => {
    const response = await dashboardApi.get<DashboardResponse>(ENDPOINTS.DASHBOARD(clientId));
    return response.data;
  },

  /**
   * Récupération du dashboard avec calculs enrichis
   */
  getEnrichedDashboard: async (clientId: number): Promise<EnrichedDashboardResponse> => {
    const dashboard = await dashboardService.getDashboard(clientId);
    
    // Calcul du solde total
    const totalBalance = dashboard.accounts.reduce((sum, account) => sum + account.balance, 0);
    
    // Calcul des revenus et dépenses du mois en cours
    const now = new Date();
    const startOfMonth = new Date(now.getFullYear(), now.getMonth(), 1).toISOString();
    
    const monthlyTransactions = dashboard.recentTransactions.filter(
      t => t.createdAt && t.createdAt >= startOfMonth
    );
    
    const monthlyIncome = monthlyTransactions
      .filter(t => t.type === 'DEPOSIT')
      .reduce((sum, t) => sum + t.amount, 0);
    
    const monthlyExpenses = monthlyTransactions
      .filter(t => t.type === 'WITHDRAWAL' || t.type === 'TRANSFER')
      .reduce((sum, t) => sum + t.amount, 0);
    
    return {
      ...dashboard,
      totalBalance,
      monthlyIncome,
      monthlyExpenses,
    };
  },

  /**
   * Récupération d'un relevé de compte
   */
  getAccountStatement: async (
    accountId: number,
    startDate: string,
    endDate: string
  ): Promise<StatementResponse> => {
    const response = await dashboardApi.get<StatementResponse>(ENDPOINTS.STATEMENT(accountId), {
      params: { startDate, endDate },
    });
    return response.data;
  },

  /**
   * Récupération d'un relevé de compte avec calculs enrichis
   */
  getEnrichedStatement: async (
    accountId: number,
    startDate: string,
    endDate: string
  ): Promise<EnrichedStatementResponse> => {
    const statement = await dashboardService.getAccountStatement(accountId, startDate, endDate);
    
    const totalCredits = statement.transactions
      .filter(t => t.type === 'DEPOSIT')
      .reduce((sum, t) => sum + t.amount, 0);
    
    const totalDebits = statement.transactions
      .filter(t => t.type === 'WITHDRAWAL' || t.type === 'TRANSFER')
      .reduce((sum, t) => sum + t.amount, 0);
    
    return {
      ...statement,
      totalCredits,
      totalDebits,
      balance: statement.account.balance,
    };
  },
};

export default dashboardService;

File: dashboard-composite-service/src/main/java/com/willbank/dashboard/service/DashboardService.java
package com.willbank.dashboard.service;

import com.willbank.dashboard.client.AccountServiceClient;
import com.willbank.dashboard.client.ClientServiceClient;
import com.willbank.dashboard.client.TransactionServiceClient;
import com.willbank.dashboard.dto.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class DashboardService {
    
    private final ClientServiceClient clientServiceClient;
    private final AccountServiceClient accountServiceClient;
    private final TransactionServiceClient transactionServiceClient;
    private final DashboardAnalyticsService analyticsService;
    
    public DashboardResponse getDashboard(Long clientId) {
        log.info("Fetching dashboard for client ID: {}", clientId);
        
        // Fetch client information
        ClientDTO client = clientServiceClient.getClientById(clientId);
        log.info("Client fetched: {}", client.getEmail());
        
        // Fetch all accounts for the client
        List<AccountDTO> accounts = accountServiceClient.getAccountsByClientId(clientId);
        log.info("Found {} accounts for client", accounts.size());
        
        // Fetch recent transactions for all accounts (last 10)
        List<TransactionDTO> allTransactions = new ArrayList<>();
        for (AccountDTO account : accounts) {
            try {
                List<TransactionDTO> accountTransactions = transactionServiceClient.getTransactionsByAccountId(account.getId());
                allTransactions.addAll(accountTransactions);
            } catch (Exception e) {
                log.warn("Could not fetch transactions for account {}: {}", account.getId(), e.getMessage());
            }
        }
        
        // Sort by date and get last 10
        List<TransactionDTO> recentTransactions = allTransactions.stream()
                .sorted(Comparator.comparing(TransactionDTO::getCreatedAt).reversed())
                .limit(10)
                .collect(Collectors.toList());
        
        log.info("Fetched {} recent transactions", recentTransactions.size());
        
        // Calculs enrichis
        BigDecimal totalBalance = accounts.stream()
            .map(AccountDTO::getBalance)
            .reduce(BigDecimal.ZERO, BigDecimal::add);
        
        BigDecimal monthlyIncome = allTransactions.stream()
            .filter(t -> "DEPOSIT".equals(t.getType()))
            .map(TransactionDTO::getAmount)
            .reduce(BigDecimal.ZERO, BigDecimal::add);
        
        BigDecimal monthlyExpenses = allTransactions.stream()
            .filter(t -> "WITHDRAWAL".equals(t.getType()) || "TRANSFER".equals(t.getType()))
            .map(TransactionDTO::getAmount)
            .reduce(BigDecimal.ZERO, BigDecimal::add);
        
        BigDecimal savingsRate = analyticsService.calculateSavingsRate(monthlyIncome, monthlyExpenses);
        Map<String, BigDecimal> transactionsByType = analyticsService.calculateTransactionsByType(allTransactions);
        List<BalanceHistoryDTO> balanceHistory = analyticsService.generateBalanceHistory(accounts, allTransactions);
        DashboardInsightsDTO insights = analyticsService.generateInsights(accounts, allTransactions);
        
        DashboardResponse response = new DashboardResponse();
        response.setClient(client);
        response.setAccounts(accounts);
        response.setRecentTransactions(recentTransactions);
        response.setTotalBalance(totalBalance);
        response.setMonthlyIncome(monthlyIncome);
        response.setMonthlyExpenses(monthlyExpenses);
        response.setSavingsRate(savingsRate);
        response.setTransactionsByType(transactionsByType);
        response.setBalanceHistory(balanceHistory);
        response.setInsights(insights);
        response.setLastUpdated(LocalDateTime.now());
        
        log.info("Dashboard enriched with analytics data for client {}", clientId);
        
        return response;
    }
    
    public StatementResponse getAccountStatement(Long accountId, LocalDateTime startDate, LocalDateTime endDate) {
        log.info("Fetching statement for account ID: {} from {} to {}", accountId, startDate, endDate);
        
        // Fetch account information
        AccountDTO account = accountServiceClient.getAccountById(accountId);
        log.info("Account fetched: {}", account.getAccountNumber());
        
        // Fetch transactions for the date range
        List<TransactionDTO> transactions = transactionServiceClient.getTransactionsByDateRange(accountId, startDate, endDate);
        log.info("Found {} transactions in the date range", transactions.size());
        
        StatementResponse response = new StatementResponse();
        response.setAccount(account);
        response.setStartDate(startDate);
        response.setEndDate(endDate);
        response.setTransactions(transactions);
        
        return response;
    }
}
